libs/wingc/src/type_check.rs

```diff
@@ -1730,9 +1733,23 @@ pub struct TypeChecker<'a> {
 
 	is_in_mut_json: bool,
 
+	// Stack of extra information we need to keep track of while visiting nested
+	// expressions during type checking.
+	curr_expr_info: Vec<ExprVisitInfo>,
```

Comment:
(random thought) I'm wondering if it's also possible to implement this by leveraging the fact that every `Expr` has an `id` field -- so you could store a lookup table that maps from IDs to the extra metadata structs. You can see this approach used for `type_for_expr`, `json_literal_casts`, and `type_expressions`.

(All that said I'm not sure if it would result in any less code)

Reply:
This is a slightly different case since we only need the information while we're doing the type checking. Ideally I should have added it to `VisitContext` since it's actually context information used during the AST visiting process of the type checking. But it seemed to "type-checking" specific to generalize it into the `VisitContext`. I'll keep it as is for now.